File: cocktail/cocktail.go
/**
 * Author: Sigifredo Escobar Gómez
 * Email: sigifredo89@gmail.com
 * Date: Sun, 23 Feb 2014 12:11:00 -0500
 * Algorithm: Cocktail Sort
 */


package main

import "fmt"

func cocktail(arr []int) {
    swap := true

    left := 1
    right := len(arr)-1

    for swap {
        swap = false

        for i := right; i > left; i-- {
            if arr[i-1] > arr[i] {
                arr[i] = arr[i] + arr[i-1]
                arr[i-1] = arr[i] - arr[i-1]
                arr[i] = arr[i] - arr[i-1]

                swap = true
            }
        }

        for i := left; i < right; i++ {
            if arr[i-1] > arr[i] {
                arr[i] = arr[i] + arr[i-1]
                arr[i-1] = arr[i] - arr[i-1]
                arr[i] = arr[i] - arr[i-1]

                swap = true
            }
        }
    }
}

func main() {
    var n int
    var arr []int

    fmt.Scanf("%d", &n)
    arr = make([]int, n)

    for i, _ := range arr {
        fmt.Scanf("%d", &arr[i])
    }

    cocktail(arr)

    for _, v := range arr {
        fmt.Printf("%d ", v)
    }
    fmt.Printf("\n")
}

File: cocktail/cocktail.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Author: Sigifredo Escobar Gómez
# Email: sigifredo89@gmail.com
# Date: Sun, 23 Feb 2014 12:01:38 -0500
# Algorithm: Cocktail Sort

def cocktail(arr):
    swap = False;

    left = 1;
    right = len(arr)-1;

    while True:
        swap = False;
        for i in range(left, right):
            if (arr[-i-1] > arr[-i]):
                arr[-i] = arr[-i] + arr[-i-1];
                arr[-i-1] = arr[-i] - arr[-i-1];
                arr[-i] = arr[-i] - arr[-i-1];

                swap = True;

        for i in range(left, right):
            if (arr[i-1] > arr[i]):
                arr[i] = arr[i] + arr[i-1];
                arr[i-1] = arr[i] - arr[i-1];
                arr[i] = arr[i] - arr[i-1];

                swap = True;

        if not swap:
            break

if __name__ == '__main__':
    n = int(raw_input(""))

    arr = range(n)

    for i in range(n):
        arr[i] = int(raw_input(""))

    cocktail(arr)

    for i in arr:
        print i,
